File: Data Structures/05-BT/AVL/AVL.py
class Node(object):
    """Node object"""

    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        self.height = 1

class AVL(object):
    def orderTree(self, root):
        root.height = 1 + max(self.getHeight(root.left), self.getHeight(root.right))

        balance = self.getBalance(root)

        if balance > 1 and self.key < root.left.val:
            return self.rightRotate(root)

        if balance < -1 and self.key > root.right.val:
            return self.leftRotate(root)

        if balance > 1 and self.key > root.left.val:
            root.left = self.leftRotate(root.left)
            return self.rightRotate(root)

        if balance < 1 and self.key < root.right.val:
            root.right = self.rightRotate(root.right)
            return self.leftRotate(root)
        return root

    def insert(self, root, key):
        if not root:
            return Node(key)
        elif key < root.val:
            root.left = self.insert(root.left, key)
        else:
            root.right = self.insert(root.right, key)

        return self.orderTree(root)

    def delete(self, root, key):
        if not root:
            return root
        elif key < root.val:
            root.left = self.delete(root.left, key)
        elif key > root.val:
            root.right = self.detele(root.right, key)

        else:
            if root.left is None:
                tmp = root.right
                root = None
                return tmp
            elif root.right is None:
                tmp = root.left
                root = None
                return tmp

            tmp = self.getMinValueNode(root.right)
            root.val = tmp.val
            root.right = self.delete(root.right, tmp.val)

        if root is None:
            return root

        return self.orderTree(root)


    def leftRotate(self, z):
        y = z.right
        tmp = y.left

        y.left = z
        z.right = tmp

        # Perform rotation
        y.left = z
        z.right = tmp

        # Update heights
        z.height = 1 + max(self.getHeight(z.left), self.getHeight(z.right))
        y.height = 1 + max(self.getHeight(y.left), self.getHeight(y.right))

        # Return the new root
        return y


    def rightRotate(self, z):
        y = z.left
        tmp = y.right

        y.right = z
        z.left = tmp

        # Update heights
        z.height = 1 + max(self.getHeight(z.left),
                        self.getHeight(z.right))
        y.height = 1 + max(self.getHeight(y.left),
                        self.getHeight(y.right))

        # Return the new root
        return y

    def getHeight(self, root):
        if not root:
            return 0

        return root.height

    def getBalance(self, root):
        if not root:
            return 0

        return self.getHeight(root.left) - self.getHeight(root.right)

    def getMinValueNode(self, root):
        if root is None or root.left is None:
            return root

        return self.getMinValueNode(root.left)

    def preOrder(self, root):
        if not root:
            return

        print("{0}".format(root.val))
        self.preOrder(root.left)
        self.preOrder(root.right)


myTree = AVL()
root = None

root = myTree.insert(root, 10)
root = myTree.insert(root, 20)
root = myTree.insert(root, 30)
root = myTree.insert(root, 40)
root = myTree.insert(root, 50)
root = myTree.insert(root, 25)

myTree.preOrder(root)
print()

File: Data Structures/05-BT/AVL/AVL.js
class Tree {
    constructor () {
        this.root = null;
    }
    add(value) {
        if (!this.root) {
            this.root = new Node(value);
        } else {
            this.root.add(value);
        }
    }
    toObject() {
        return this.root;
    }
}

class Node {
    constructor (value) {
        this.left = null;
        this.right = null;
        this.value = value;
        this.height = 1;
    }
    add(value) {
        // decide to go left or right
        if (value < this.value) {
            if (this.left) {
                this.left.add(value);
            } else {
                this.left = new Node(value);
            }
            if (!this.right || this.right.height < this.left.height) {
                this.height = this.left.height + 1;
            }
        } else {
            if (this.right) {
                this.right.add(value);
            } else {
                this.right = new Node(value);
            }
            if (!this.left || this.right.height > this.left.height) {
                this.height = this.right.height + 1;
            }
        }
        this.balance();
    }
    balance() {
        const rightHeight = this.right ? this.right.height : 0;
        const leftHeight = this.left ? this.left.height : 0;

        if (leftHeight > rightHeight + 1) {
            const leftRightHeight = this.left.right ? this.left.right.height : 0;
            const leftLeftHeight = this.left.left ? this.left.left : 0;
        
            if (leftRightHeight > leftLeftHeight) {
                this.left.rotateRR();
            }
            this.rotateLL();
        } else if (rightHeight > leftHeight + 1) {
            const rightRightHeight = this.right.right ? this.right.right.height : 0;
            const rightLeftHeight = this.right.left? this.right.left.height : 0;

            if (rightLeftHeight > rightRightHeight) {
                this.right.rotateLL();
            }
            this.rotateRR();    
        }
    }
    rotateRR() {
        const valueBefore = this.value;
        const leftBefore = this.left
        this.value = this.right.value;
        this.left = this.right;
        this.right = this.right.right;
        this.left.right = this.left.left;
        this.left.left = leftBefore;
        this.left.value = valueBefore;

        this.left.updateInNewLocation();
        this.updateInNewLocation();
    }
    rotateLL() {
        const valueBefore = this.value;
        const rightBefore = this.right;
        this.value = this.left.value;
        this.right = this.left;
        this.left = this.left.left;
        this.right.left = this.right.right;
        this.right.right = rightBefore;
        this.right.value = valueBefore;

        this.right.updateInNewLocation();
        this.updateInNewLocation();
    }
    updateInNewLocation() {
        if (!this.right && !this.left) {
            this.height = 1;
        } else if (!this.right || (this.left && this.right.height < this.left.height)) {
            this.height = this.left.height + 1;
        } else {
            this.height = this.right.height + 1;
        }
    }
}

/*
5
 \
  8

-> Currently valid AVL tree
-> .add called with 9

5 - node A
 \
  8 - node B
   \
    9 - node C

(on the way up from the recursion)
-> check balance of node C: left height is 0, right height is 0, balanced
-> check balance of node B: left height is 0, right height is 1, balanced
-> check balance of node A: left height is 0, right height is 2
unbalanced, right heavy, child is right heavy

-> perform right rotation
-> swap the values of nodes A and B
-> make node B the left child of node A
-> make node C the right child of node A
-> move node B's right child to its left child
(in this case they're both null)
-> make node A's _original_ left child
(which was null in this case) the left child of node B
-> update the heights of all the nodes involved

      8 - node A
    /   \
   5     9
node B   node C
*/

/* 
55 - node A
 \
  8 - node B
 /
7 - node C

[ ... previous steps ]
-> check balance of node A: left height is 0, right height is 2
unbalanced - right heavy, child is left heavy
-> perform left rotation on left heavy right child node B

5 - node A
 \
  7 - node B
   \
    8 - node C

-> now perform right rotation on node A

      7 - node A
    /   \
   5     8
node B   node C
*/